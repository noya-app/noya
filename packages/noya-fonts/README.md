# Noya Fonts

A utility for managing fonts
