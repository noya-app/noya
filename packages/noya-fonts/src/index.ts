import { Brand } from 'noya-utils';

export * from './fontTraits';
export * from './fontDescriptor';
export * from './fontManager';
export * from './fontWeight';

export function formatFontFamilyId(fontFamily: string) {
  return fontFamily.toLowerCase().replace(/[ _-]/g, '');
}

export type FontFamilyId = Brand<string, 'fontFamily'>;
