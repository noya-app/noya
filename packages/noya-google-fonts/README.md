# Noya Google Fonts

Utilities for working with Google Fonts. We include a pre-downloaded list of
fonts so that they can be ready as soon as possible.
