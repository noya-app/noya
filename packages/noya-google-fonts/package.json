{
  "name": "noya-google-fonts",
  "version": "1.0.0",
  "main": "src/index.ts",
  "license": "UNLICENSED"
}
