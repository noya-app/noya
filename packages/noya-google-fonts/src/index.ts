const webfontList: WebfontList = require('./fonts.json');
const fonts = webfontList.items;
const fontMap = fonts.reduce((result: Record<string, WebfontFamily>, item) => {
  result[item.family.toLowerCase()] = item;
  return result;
}, {});

type FontCategory =
  | 'sans-serif'
  | 'serif'
  | 'display'
  | 'handwriting'
  | 'monospace';

type FontSubset =
  | 'latin'
  | 'latin-ext'
  | 'sinhala'
  | 'greek'
  | 'kannada'
  | 'telugu'
  | 'vietnamese'
  | 'hebrew'
  | 'cyrillic'
  | 'cyrillic-ext'
  | 'greek-ext'
  | 'arabic'
  | 'devanagari'
  | 'khmer'
  | 'tamil'
  | 'thai'
  | 'bengali'
  | 'gujarati'
  | 'oriya'
  | 'malayalam'
  | 'gurmukhi'
  | 'korean'
  | 'japanese'
  | 'tibetan'
  | 'chinese-simplified'
  | 'chinese-hongkong'
  | 'chinese-traditional'
  | 'myanmar';

export type RegularFontVariant =
  | '100'
  | '200'
  | '300'
  | 'regular'
  | '500'
  | '600'
  | '700'
  | '800'
  | '900';

export type ItalicFontVariant =
  | '100italic'
  | '200italic'
  | '300italic'
  | 'italic'
  | '500italic'
  | '600italic'
  | '700italic'
  | '800italic'
  | '900italic';

export type FontVariant = RegularFontVariant | ItalicFontVariant;

interface WebfontFamily {
  family: string;
  variants: FontVariant[];
  subsets: FontSubset[];
  version: string;
  lastModified: string;
  files: Record<FontVariant, string>;
  category: FontCategory;
  kind: 'webfonts#webfont';
}

interface WebfontList {
  items: WebfontFamily[];
  kind: 'webfonts#webfontList';
}

export function getFontFamilyList() {
  return fonts.map((font) => font.family);
}

export function hasFontFamily(fontFamily: string) {
  return getKey(fontFamily) in fontMap;
}

export type FontVariantCollection = {
  regular: RegularFontVariant[];
  italic: ItalicFontVariant[];
};

function isItalicVariant(variant: FontVariant): variant is ItalicFontVariant {
  return variant.includes('italic');
}

function isRegularVariant(variant: FontVariant): variant is RegularFontVariant {
  return !variant.includes('italic');
}

export function getFontVariants(fontFamily: string) {
  return fontMap[getKey(fontFamily)].variants;
}

export function getFontVariantCollection(
  fontFamily: string,
): FontVariantCollection {
  const variants = fontMap[getKey(fontFamily)].variants;
  const regular = variants.filter(isRegularVariant);
  const italic = variants.filter(isItalicVariant);

  return { regular, italic };
}

export function getFontFile(fontFamily: string, variant: FontVariant) {
  return fontMap[getKey(fontFamily)].files[variant];
}

function getKey(fontFamily: string) {
  return fontFamily.toLowerCase();
}

type FontWeight =
  | 'ultralight'
  | 'thin'
  | 'light'
  | 'regular'
  | 'medium'
  | 'semibold'
  | 'bold'
  | 'heavy'
  | 'black';

export function getFontVariantWeight(variant: FontVariant) {
  switch (variant) {
    case '100':
    case '100italic':
      return 'ultralight';
    case '200':
    case '200italic':
      return 'thin';
    case '300':
    case '300italic':
      return 'light';
    case 'regular':
    case 'italic':
      return 'regular';
    case '500':
    case '500italic':
      return 'medium';
    case '600':
    case '600italic':
      return 'semibold';
    case '700':
    case '700italic':
      return 'bold';
    case '800':
    case '800italic':
      return 'heavy';
    case '900':
    case '900italic':
      return 'black';
  }
}

export function isValidFontVariant(string: string): string is FontVariant {
  switch (string) {
    case '100':
    case '100italic':
    case '200':
    case '200italic':
    case '300':
    case '300italic':
    case 'regular':
    case 'italic':
    case '500':
    case '500italic':
    case '600':
    case '600italic':
    case '700':
    case '700italic':
    case '800':
    case '800italic':
    case '900':
    case '900italic':
      return true;
    default:
      return false;
  }
}

export function decodeFontVariant(
  variant: FontVariant,
): {
  weight: FontWeight;
  variantName: 'regular' | 'italic';
} {
  return {
    weight: getFontVariantWeight(variant),
    variantName: isItalicVariant(variant) ? 'italic' : 'regular',
  };
}
