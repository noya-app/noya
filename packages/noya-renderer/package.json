{
  "name": "noya-renderer",
  "version": "1.0.0",
  "main": "src/index.ts",
  "license": "UNLICENSED",
  "dependencies": {
    "@sketch-hq/sketch-file-format-ts": "^6.0.3",
    "canvaskit-wasm": "0.25.0",
    "uuid": "^8.3.2"
  },
  "devDependencies": {
    "@types/uuid": "^8.3.0",
    "canvas": "^2.6.1"
  }
}
