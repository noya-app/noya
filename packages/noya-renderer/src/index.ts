import type { FontMgr, Paint, PaintStyle } from 'canvaskit';
import { CanvasKitInit } from 'canvaskit';
import { v4 as uuid } from 'uuid';
import { Context } from './context';
import * as Primitives from './primitives';
export { default as LayerPreview } from './components/LayerPreview';
export { default as SketchArtboard } from './components/layers/SketchArtboard';
export { default as SketchGroup } from './components/layers/SketchGroup';
export { default as SketchLayer } from './components/layers/SketchLayer';
export { default as SketchFileRenderer } from './components/SketchFileRenderer';
export * from './ComponentsContext';
export { ImageCacheProvider } from './ImageCache';
export type { Context };
export { uuid, Primitives };

export type SimpleTextDecoration = Primitives.SimpleTextDecoration;

export let fontManager: FontMgr;

declare module 'canvaskit' {
  interface Paint {
    style?: PaintStyle;
  }
}

export async function load() {
  const [CanvasKit, fontBuffer] = await Promise.all([
    CanvasKitInit({
      locateFile: (file: string) => '/wasm/' + file,
    }),
    fetch(
      'https://storage.googleapis.com/skia-cdn/google-web-fonts/Roboto-Regular.ttf',
    ).then((resp) => resp.arrayBuffer()),
  ]);

  fontManager = CanvasKit.FontMgr.FromData(fontBuffer)!;

  const _setStyle = CanvasKit.Paint.prototype.setStyle;

  CanvasKit.Paint.prototype.setStyle = function (
    this: Paint,
    paintStyle: PaintStyle,
  ) {
    this.style = paintStyle;
    _setStyle.call(this, paintStyle);
  };

  return CanvasKit;
}
