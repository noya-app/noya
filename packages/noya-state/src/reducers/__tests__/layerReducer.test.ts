import type { CanvasKit as CanvasKitType } from 'canvaskit';
import { FontManager } from 'noya-fonts';
import { GoogleFontProvider } from 'noya-google-fonts';
import { loadCanvasKit } from 'noya-renderer';
import { debugDescription, SketchModel } from 'noya-sketch-model';
import {
  createInitialState,
  createSketchFile,
  Layers,
  Selectors,
} from 'noya-state';
import {
  ApplicationReducerContext,
  ApplicationState,
} from '../applicationReducer';
import { layerReducer } from '../layerReducer';

let CanvasKit: CanvasKitType;
let context: ApplicationReducerContext;

beforeAll(async () => {
  CanvasKit = await loadCanvasKit();
  const typefaceFontProvider = CanvasKit.TypefaceFontProvider.Make();
  context = {
    canvasInsets: { top: 0, right: 0, bottom: 0, left: 0 },
    canvasSize: { width: 1000, height: 1000 },
    fontManager: {
      ...new FontManager(GoogleFontProvider),
      getTypefaceFontProvider: () => typefaceFontProvider,
    },
  };
});

const getPageLayersLength = (state: ApplicationState) =>
  Selectors.getCurrentPage(state).layers.length;

const rectangle = SketchModel.rectangle({
  frame: SketchModel.rect({
    width: 100,
    height: 100,
  }),
});
const oval = SketchModel.oval({
  frame: SketchModel.rect({
    x: 50,
    y: 50,
    width: 100,
    height: 100,
  }),
});
const artboard = SketchModel.artboard();
const text = SketchModel.text();

describe('moveLayer', () => {
  test('move one inside', () => {
    const group = SketchModel.group();

    const state = createInitialState(
      createSketchFile(SketchModel.page({ layers: [rectangle, group] })),
    );

    const updated = layerReducer(
      state,
      ['moveLayer', rectangle.do_objectID, group.do_objectID, 'inside'],
      context,
    );

    expect(
      debugDescription([
        Selectors.getCurrentPage(state),
        Selectors.getCurrentPage(updated),
      ]),
    ).toMatchSnapshot();
  });

  test('move one above', () => {
    const group = SketchModel.group();

    const state = createInitialState(
      createSketchFile(SketchModel.page({ layers: [rectangle, group] })),
    );

    const updated = layerReducer(
      state,
      ['moveLayer', rectangle.do_objectID, group.do_objectID, 'above'],
      context,
    );

    expect(
      debugDescription([
        Selectors.getCurrentPage(state),
        Selectors.getCurrentPage(updated),
      ]),
    ).toMatchSnapshot();
  });

  test('move one below', () => {
    const group = SketchModel.group();

    const state = createInitialState(
      createSketchFile(SketchModel.page({ layers: [rectangle, group] })),
    );

    const updated = layerReducer(
      state,
      ['moveLayer', group.do_objectID, rectangle.do_objectID, 'below'],
      context,
    );

    expect(
      debugDescription([
        Selectors.getCurrentPage(state),
        Selectors.getCurrentPage(updated),
      ]),
    ).toMatchSnapshot();
  });

  test('move already in correct place', () => {
    const group = SketchModel.group();

    const state = createInitialState(
      createSketchFile(SketchModel.page({ layers: [rectangle, group] })),
    );

    expect(state).toEqual(
      layerReducer(
        state,
        ['moveLayer', rectangle.do_objectID, group.do_objectID, 'below'],
        context,
      ),
    );

    expect(state).toEqual(
      layerReducer(
        state,
        ['moveLayer', group.do_objectID, rectangle.do_objectID, 'above'],
        context,
      ),
    );
  });

  test('move multiple inside', () => {
    const group = SketchModel.group();

    const state = createInitialState(
      createSketchFile(SketchModel.page({ layers: [rectangle, oval, group] })),
    );

    const updated = layerReducer(
      state,
      [
        'moveLayer',
        [rectangle.do_objectID, oval.do_objectID],
        group.do_objectID,
        'inside',
      ],
      context,
    );

    expect(
      debugDescription([
        Selectors.getCurrentPage(state),
        Selectors.getCurrentPage(updated),
      ]),
    ).toMatchSnapshot();
  });
});

describe('deleteLayer', () => {
  test('delete one', () => {
    const state = createInitialState(
      createSketchFile(SketchModel.page({ layers: [rectangle] })),
    );

    expect(getPageLayersLength(state)).toEqual(1);

    const updated = layerReducer(
      state,
      ['deleteLayer', rectangle.do_objectID],
      context,
    );

    expect(getPageLayersLength(updated)).toEqual(0);
  });

  const group = SketchModel.group({ layers: [rectangle] });

  describe('delete multiple', () => {
    const state = createInitialState(
      createSketchFile(SketchModel.page({ layers: [group, oval] })),
    );

    test('within same parent', () => {
      const updated = layerReducer(
        state,
        ['deleteLayer', [group.do_objectID, oval.do_objectID]],
        context,
      );

      expect(
        debugDescription([
          Selectors.getCurrentPage(state),
          Selectors.getCurrentPage(updated),
        ]),
      ).toMatchSnapshot();
    });

    test('within different parents', () => {
      const updated = layerReducer(
        state,
        ['deleteLayer', [rectangle.do_objectID, oval.do_objectID]],
        context,
      );

      expect(
        debugDescription([
          Selectors.getCurrentPage(state),
          Selectors.getCurrentPage(updated),
        ]),
      ).toMatchSnapshot();
    });

    // Rectangle is a child of Group, so this should be an invalid selection
    test('bad nesting', () => {
      const updated = layerReducer(
        state,
        ['deleteLayer', [group.do_objectID, rectangle.do_objectID]],
        context,
      );

      expect(
        debugDescription([
          Selectors.getCurrentPage(state),
          Selectors.getCurrentPage(updated),
        ]),
      ).toMatchSnapshot();
    });
  });
});

describe('duplicateLayer', () => {
  test('duplicate one', () => {
    const state = createInitialState(
      createSketchFile(SketchModel.page({ layers: [rectangle] })),
    );

    const updated = layerReducer(
      state,
      ['duplicateLayer', [rectangle.do_objectID]],
      context,
    );

    expect(
      debugDescription([
        Selectors.getCurrentPage(state),
        Selectors.getCurrentPage(updated),
      ]),
    ).toMatchSnapshot();
  });

  test('duplicate multiple', () => {
    const state = createInitialState(
      createSketchFile(SketchModel.page({ layers: [rectangle, oval] })),
    );

    const updated = layerReducer(
      state,
      ['duplicateLayer', [rectangle.do_objectID, oval.do_objectID]],
      context,
    );

    expect(
      debugDescription([
        Selectors.getCurrentPage(state),
        Selectors.getCurrentPage(updated),
      ]),
    ).toMatchSnapshot();
  });

  test('duplicate multiple with gap', () => {
    const state = createInitialState(
      createSketchFile(
        SketchModel.page({ layers: [rectangle, oval, artboard, text] }),
      ),
    );

    const updated = layerReducer(
      state,
      ['duplicateLayer', [rectangle.do_objectID, artboard.do_objectID]],
      context,
    );

    expect(
      debugDescription([
        Selectors.getCurrentPage(state),
        Selectors.getCurrentPage(updated),
      ]),
    ).toMatchSnapshot();
  });

  test('duplicate multiple in different parents', () => {
    const artboard = SketchModel.artboard({ layers: [text] });

    const state = createInitialState(
      createSketchFile(SketchModel.page({ layers: [rectangle, artboard] })),
    );

    const updated = layerReducer(
      state,
      ['duplicateLayer', [rectangle.do_objectID, text.do_objectID]],
      context,
    );

    expect(
      debugDescription([
        Selectors.getCurrentPage(state),
        Selectors.getCurrentPage(updated),
      ]),
    ).toMatchSnapshot();
  });
});

describe('grouping', () => {
  test('Group & ungroup layers', () => {
    const rectangle = SketchModel.rectangle({
      frame: SketchModel.rect({
        x: 100,
        y: 100,
        width: 100,
        height: 100,
      }),
    });

    const oval = SketchModel.oval({
      frame: SketchModel.rect({
        x: 250,
        y: 250,
        width: 100,
        height: 100,
      }),
    });

    const state = createInitialState(
      createSketchFile(
        SketchModel.page({
          layers: [
            SketchModel.rectangle({ name: 'Bottom' }),
            rectangle,
            oval,
            SketchModel.rectangle({ name: 'Top' }),
          ],
        }),
      ),
    );

    const groupedState = layerReducer(
      state,
      ['groupLayers', [rectangle.do_objectID, oval.do_objectID]],
      context,
    );

    expect(
      debugDescription([
        Selectors.getCurrentPage(state),
        Selectors.getCurrentPage(groupedState),
      ]),
    ).toMatchSnapshot();

    const group = Layers.find(
      Selectors.getCurrentPage(groupedState),
      (layer) => layer._class === 'group',
    )!;

    expect(groupedState.selectedLayerIds).toEqual([group.do_objectID]);

    const ungroupedState = layerReducer(
      groupedState,
      ['ungroupLayers', [group.do_objectID]],
      context,
    );

    expect(
      debugDescription([
        Selectors.getCurrentPage(groupedState),
        Selectors.getCurrentPage(ungroupedState),
      ]),
    ).toMatchSnapshot();

    expect(ungroupedState.selectedLayerIds).toEqual([
      rectangle.do_objectID,
      oval.do_objectID,
    ]);
  });

  test('Group & ungroup nested layers', () => {
    const rectangle = SketchModel.rectangle({
      frame: SketchModel.rect({
        x: 100,
        y: 100,
        width: 100,
        height: 100,
      }),
    });

    const oval = SketchModel.oval({
      frame: SketchModel.rect({
        x: 250,
        y: 250,
        width: 100,
        height: 100,
      }),
    });

    const state = createInitialState(
      createSketchFile(
        SketchModel.page({
          layers: [
            SketchModel.artboard({
              frame: SketchModel.rect({
                x: 100,
                y: 100,
                width: 300,
                height: 300,
              }),
              layers: [
                SketchModel.rectangle({ name: 'Bottom' }),
                rectangle,
                oval,
                SketchModel.rectangle({ name: 'Top' }),
              ],
            }),
          ],
        }),
      ),
    );

    const groupedState = layerReducer(
      state,
      ['groupLayers', [rectangle.do_objectID, oval.do_objectID]],
      context,
    );

    expect(
      debugDescription([
        Selectors.getCurrentPage(state),
        Selectors.getCurrentPage(groupedState),
      ]),
    ).toMatchSnapshot();

    const group = Layers.find(
      Selectors.getCurrentPage(groupedState),
      (layer) => layer._class === 'group',
    )!;

    expect(groupedState.selectedLayerIds).toEqual([group.do_objectID]);

    const ungroupedState = layerReducer(
      groupedState,
      ['ungroupLayers', [group.do_objectID]],
      context,
    );

    expect(
      debugDescription([
        Selectors.getCurrentPage(groupedState),
        Selectors.getCurrentPage(ungroupedState),
      ]),
    ).toMatchSnapshot();

    expect(ungroupedState.selectedLayerIds).toEqual([
      rectangle.do_objectID,
      oval.do_objectID,
    ]);
  });

  test('Ungroup multiple groups', () => {
    const rect1 = SketchModel.rectangle({ name: '1' });
    const rect2 = SketchModel.rectangle({ name: '2' });
    const rect3 = SketchModel.rectangle({ name: '3' });
    const rect4 = SketchModel.rectangle({ name: '4' });

    const group1 = SketchModel.group({
      layers: [rect1, rect2],
    });
    const group2 = SketchModel.group({
      layers: [rect3, rect4],
    });

    const state = createInitialState(
      createSketchFile(
        SketchModel.page({
          layers: [
            rectangle,
            group1,
            SketchModel.rectangle({ name: 'Middle' }),
            group2,
            oval,
          ],
        }),
      ),
    );

    state.selectedLayerIds = [
      rectangle.do_objectID,
      oval.do_objectID,
      group1.do_objectID,
      group2.do_objectID,
    ];

    const updated = layerReducer(
      state,
      ['ungroupLayers', state.selectedLayerIds],
      context,
    );

    expect(
      debugDescription([
        Selectors.getCurrentPage(state),
        Selectors.getCurrentPage(updated),
      ]),
    ).toMatchSnapshot();

    expect(updated.selectedLayerIds).toEqual([
      rectangle.do_objectID,
      oval.do_objectID,
      rect1.do_objectID,
      rect2.do_objectID,
      rect3.do_objectID,
      rect4.do_objectID,
    ]);
  });
});

describe('addLayer', () => {
  test('selected layer is a group', () => {
    const state = createInitialState(
      createSketchFile(SketchModel.page({ layers: [rectangle, oval] })),
    );

    const groupedState = layerReducer(
      state,
      ['groupLayers', [rectangle.do_objectID, oval.do_objectID]],
      context,
    );

    const updated = layerReducer(
      groupedState,
      ['addLayer', rectangle],
      context,
    );

    expect(
      debugDescription([
        Selectors.getCurrentPage(groupedState),
        Selectors.getCurrentPage(updated),
      ]),
    ).toMatchSnapshot();
  });
  test('original layer is in parent layer', () => {
    const state = createInitialState(
      createSketchFile(SketchModel.page({ layers: [rectangle, oval] })),
    );

    const updated = layerReducer(state, ['addLayer', rectangle], context);

    expect(
      debugDescription([
        Selectors.getCurrentPage(state),
        Selectors.getCurrentPage(updated),
      ]),
    ).toMatchSnapshot();
  });
  test('some layer is selected', () => {
    const rectangleTwo = SketchModel.rectangle({
      frame: SketchModel.rect({
        x: 100,
        y: 100,
        width: 100,
        height: 100,
      }),
    });

    const state = createInitialState(
      createSketchFile(
        SketchModel.page({ layers: [oval, text, rectangleTwo] }),
      ),
    );

    const stateWithLayerSelected = layerReducer(
      state,
      ['selectLayer', oval.do_objectID],
      context,
    );

    const updated = layerReducer(
      stateWithLayerSelected,
      ['addLayer', rectangle],
      context,
    );

    expect(
      debugDescription([
        Selectors.getCurrentPage(state),
        Selectors.getCurrentPage(updated),
      ]),
    ).toMatchSnapshot();
  });
  test('no layer is selected', () => {
    const state = createInitialState(
      createSketchFile(SketchModel.page({ layers: [oval] })),
    );

    const updated = layerReducer(state, ['addLayer', rectangle], context);

    expect(
      debugDescription([
        Selectors.getCurrentPage(state),
        Selectors.getCurrentPage(updated),
      ]),
    ).toMatchSnapshot();
  });
});
