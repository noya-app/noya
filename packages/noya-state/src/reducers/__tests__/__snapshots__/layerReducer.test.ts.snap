// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`deleteLayer delete multiple bad nesting 1`] = `
"<Page> Page { x: 0, y: 0, w: 0, h: 0 }
├── <Group> Group { x: 0, y: 0, w: 0, h: 0 }
│   └── <Rectangle> Rectangle { x: 0, y: 0, w: 100, h: 100 }
└── <Oval> Oval { x: 50, y: 50, w: 100, h: 100 }

<Page> Page { x: 0, y: 0, w: 0, h: 0 }
└── <Oval> Oval { x: 50, y: 50, w: 100, h: 100 }"
`;

exports[`deleteLayer delete multiple within different parents 1`] = `
"<Page> Page { x: 0, y: 0, w: 0, h: 0 }
├── <Group> Group { x: 0, y: 0, w: 0, h: 0 }
│   └── <Rectangle> Rectangle { x: 0, y: 0, w: 100, h: 100 }
└── <Oval> Oval { x: 50, y: 50, w: 100, h: 100 }

<Page> Page { x: 0, y: 0, w: 0, h: 0 }
└── <Group> Group { x: 0, y: 0, w: 0, h: 0 }"
`;

exports[`deleteLayer delete multiple within same parent 1`] = `
"<Page> Page { x: 0, y: 0, w: 0, h: 0 }
├── <Group> Group { x: 0, y: 0, w: 0, h: 0 }
│   └── <Rectangle> Rectangle { x: 0, y: 0, w: 100, h: 100 }
└── <Oval> Oval { x: 50, y: 50, w: 100, h: 100 }

<Page> Page { x: 0, y: 0, w: 0, h: 0 }"
`;

exports[`duplicateLayer duplicate multiple 1`] = `
"<Page> Page { x: 0, y: 0, w: 0, h: 0 }
├── <Rectangle> Rectangle { x: 0, y: 0, w: 100, h: 100 }
└── <Oval> Oval { x: 50, y: 50, w: 100, h: 100 }

<Page> Page { x: 0, y: 0, w: 0, h: 0 }
├── <Rectangle> Rectangle { x: 0, y: 0, w: 100, h: 100 }
├── <Oval> Oval { x: 50, y: 50, w: 100, h: 100 }
├── <Rectangle> Rectangle Copy { x: 0, y: 0, w: 100, h: 100 }
└── <Oval> Oval Copy { x: 50, y: 50, w: 100, h: 100 }"
`;

exports[`duplicateLayer duplicate multiple in different parents 1`] = `
"<Page> Page { x: 0, y: 0, w: 0, h: 0 }
├── <Rectangle> Rectangle { x: 0, y: 0, w: 100, h: 100 }
└── <Artboard> Artboard { x: 0, y: 0, w: 0, h: 0 }
    └── <Text> Text { x: 0, y: 0, w: 0, h: 0 }

<Page> Page { x: 0, y: 0, w: 0, h: 0 }
├── <Rectangle> Rectangle { x: 0, y: 0, w: 100, h: 100 }
├── <Rectangle> Rectangle Copy { x: 0, y: 0, w: 100, h: 100 }
└── <Artboard> Artboard { x: 0, y: 0, w: 0, h: 0 }
    ├── <Text> Text { x: 0, y: 0, w: 0, h: 0 }
    └── <Text> Text Copy { x: 0, y: 0, w: 0, h: 0 }"
`;

exports[`duplicateLayer duplicate multiple with gap 1`] = `
"<Page> Page { x: 0, y: 0, w: 0, h: 0 }
├── <Rectangle> Rectangle { x: 0, y: 0, w: 100, h: 100 }
├── <Oval> Oval { x: 50, y: 50, w: 100, h: 100 }
├── <Artboard> Artboard { x: 0, y: 0, w: 0, h: 0 }
└── <Text> Text { x: 0, y: 0, w: 0, h: 0 }

<Page> Page { x: 0, y: 0, w: 0, h: 0 }
├── <Rectangle> Rectangle { x: 0, y: 0, w: 100, h: 100 }
├── <Oval> Oval { x: 50, y: 50, w: 100, h: 100 }
├── <Artboard> Artboard { x: 0, y: 0, w: 0, h: 0 }
├── <Rectangle> Rectangle Copy { x: 0, y: 0, w: 100, h: 100 }
├── <Artboard> Artboard Copy { x: 100, y: 0, w: 0, h: 0 }
└── <Text> Text { x: 0, y: 0, w: 0, h: 0 }"
`;

exports[`duplicateLayer duplicate one 1`] = `
"<Page> Page { x: 0, y: 0, w: 0, h: 0 }
└── <Rectangle> Rectangle { x: 0, y: 0, w: 100, h: 100 }

<Page> Page { x: 0, y: 0, w: 0, h: 0 }
├── <Rectangle> Rectangle { x: 0, y: 0, w: 100, h: 100 }
└── <Rectangle> Rectangle Copy { x: 0, y: 0, w: 100, h: 100 }"
`;

exports[`grouping Group & ungroup layers 1`] = `
"<Page> Page { x: 0, y: 0, w: 0, h: 0 }
├── <Rectangle> Bottom { x: 0, y: 0, w: 0, h: 0 }
├── <Rectangle> Rectangle { x: 100, y: 100, w: 100, h: 100 }
├── <Oval> Oval { x: 250, y: 250, w: 100, h: 100 }
└── <Rectangle> Top { x: 0, y: 0, w: 0, h: 0 }

<Page> Page { x: 0, y: 0, w: 0, h: 0 }
├── <Rectangle> Bottom { x: 0, y: 0, w: 0, h: 0 }
├── <Group> Group { x: 100, y: 100, w: 250, h: 250 }
│   ├── <Rectangle> Rectangle { x: 0, y: 0, w: 100, h: 100 }
│   └── <Oval> Oval { x: 150, y: 150, w: 100, h: 100 }
└── <Rectangle> Top { x: 0, y: 0, w: 0, h: 0 }"
`;

exports[`grouping Group & ungroup layers 2`] = `
"<Page> Page { x: 0, y: 0, w: 0, h: 0 }
├── <Rectangle> Bottom { x: 0, y: 0, w: 0, h: 0 }
├── <Group> Group { x: 100, y: 100, w: 250, h: 250 }
│   ├── <Rectangle> Rectangle { x: 0, y: 0, w: 100, h: 100 }
│   └── <Oval> Oval { x: 150, y: 150, w: 100, h: 100 }
└── <Rectangle> Top { x: 0, y: 0, w: 0, h: 0 }

<Page> Page { x: 0, y: 0, w: 0, h: 0 }
├── <Rectangle> Bottom { x: 0, y: 0, w: 0, h: 0 }
├── <Rectangle> Rectangle { x: 100, y: 100, w: 100, h: 100 }
├── <Oval> Oval { x: 250, y: 250, w: 100, h: 100 }
└── <Rectangle> Top { x: 0, y: 0, w: 0, h: 0 }"
`;

exports[`grouping Group & ungroup nested layers 1`] = `
"<Page> Page { x: 0, y: 0, w: 0, h: 0 }
└── <Artboard> Artboard { x: 100, y: 100, w: 300, h: 300 }
    ├── <Rectangle> Bottom { x: 0, y: 0, w: 0, h: 0 }
    ├── <Rectangle> Rectangle { x: 100, y: 100, w: 100, h: 100 }
    ├── <Oval> Oval { x: 250, y: 250, w: 100, h: 100 }
    └── <Rectangle> Top { x: 0, y: 0, w: 0, h: 0 }

<Page> Page { x: 0, y: 0, w: 0, h: 0 }
└── <Artboard> Artboard { x: 100, y: 100, w: 300, h: 300 }
    ├── <Rectangle> Bottom { x: 0, y: 0, w: 0, h: 0 }
    ├── <Group> Group { x: 100, y: 100, w: 250, h: 250 }
    │   ├── <Rectangle> Rectangle { x: 0, y: 0, w: 100, h: 100 }
    │   └── <Oval> Oval { x: 150, y: 150, w: 100, h: 100 }
    └── <Rectangle> Top { x: 0, y: 0, w: 0, h: 0 }"
`;

exports[`grouping Group & ungroup nested layers 2`] = `
"<Page> Page { x: 0, y: 0, w: 0, h: 0 }
└── <Artboard> Artboard { x: 100, y: 100, w: 300, h: 300 }
    ├── <Rectangle> Bottom { x: 0, y: 0, w: 0, h: 0 }
    ├── <Group> Group { x: 100, y: 100, w: 250, h: 250 }
    │   ├── <Rectangle> Rectangle { x: 0, y: 0, w: 100, h: 100 }
    │   └── <Oval> Oval { x: 150, y: 150, w: 100, h: 100 }
    └── <Rectangle> Top { x: 0, y: 0, w: 0, h: 0 }

<Page> Page { x: 0, y: 0, w: 0, h: 0 }
└── <Artboard> Artboard { x: 100, y: 100, w: 300, h: 300 }
    ├── <Rectangle> Bottom { x: 0, y: 0, w: 0, h: 0 }
    ├── <Rectangle> Rectangle { x: 100, y: 100, w: 100, h: 100 }
    ├── <Oval> Oval { x: 250, y: 250, w: 100, h: 100 }
    └── <Rectangle> Top { x: 0, y: 0, w: 0, h: 0 }"
`;

exports[`moveLayer move multiple inside 1`] = `
"<Page> Page { x: 0, y: 0, w: 0, h: 0 }
├── <Rectangle> Rectangle { x: 0, y: 0, w: 100, h: 100 }
├── <Oval> Oval { x: 50, y: 50, w: 100, h: 100 }
└── <Group> Group { x: 0, y: 0, w: 0, h: 0 }

<Page> Page { x: 0, y: 0, w: 0, h: 0 }
└── <Group> Group { x: 0, y: 0, w: 150, h: 150 }
    ├── <Rectangle> Rectangle { x: 0, y: 0, w: 100, h: 100 }
    └── <Oval> Oval { x: 50, y: 50, w: 100, h: 100 }"
`;

exports[`moveLayer move one above 1`] = `
"<Page> Page { x: 0, y: 0, w: 0, h: 0 }
├── <Rectangle> Rectangle { x: 0, y: 0, w: 100, h: 100 }
└── <Group> Group { x: 0, y: 0, w: 0, h: 0 }

<Page> Page { x: 0, y: 0, w: 0, h: 0 }
├── <Group> Group { x: 0, y: 0, w: 0, h: 0 }
└── <Rectangle> Rectangle { x: 0, y: 0, w: 100, h: 100 }"
`;

exports[`moveLayer move one below 1`] = `
"<Page> Page { x: 0, y: 0, w: 0, h: 0 }
├── <Rectangle> Rectangle { x: 0, y: 0, w: 100, h: 100 }
└── <Group> Group { x: 0, y: 0, w: 0, h: 0 }

<Page> Page { x: 0, y: 0, w: 0, h: 0 }
├── <Group> Group { x: 0, y: 0, w: 0, h: 0 }
└── <Rectangle> Rectangle { x: 0, y: 0, w: 100, h: 100 }"
`;

exports[`moveLayer move one inside 1`] = `
"<Page> Page { x: 0, y: 0, w: 0, h: 0 }
├── <Rectangle> Rectangle { x: 0, y: 0, w: 100, h: 100 }
└── <Group> Group { x: 0, y: 0, w: 0, h: 0 }

<Page> Page { x: 0, y: 0, w: 0, h: 0 }
└── <Group> Group { x: 0, y: 0, w: 100, h: 100 }
    └── <Rectangle> Rectangle { x: 0, y: 0, w: 100, h: 100 }"
`;
